fiftyone/core/models.py

```diff
@@ -0,0 +1,687 @@
+            _merge_config(model_config_dict["config"]["config"], kwargs)
+        else:
+            _merge_config(model_config_dict["config"], kwargs)
+
+    # Load model config
+    config = ModelConfig.from_dict(model_config_dict)
+
+    #
+    # Inject model path
+    #
+    # Models must be implemented in one of the following ways in order for
+    # us to know how to inject ``model_path``:
+    #
+    # (1) Their config implements ``eta.core.learning.HasPublishedModel``
+    #
+    # (2) Their config is an ``fiftyone.core.eta_utils.ETAModelConfig`` whose
+    #     embedded config implements ``eta.core.learning.HasPublishedModel``
+    #
+    if model_path:
+        if isinstance(config.config, etal.HasPublishedModel):
+            config.config.model_name = None
+            config.config.model_path = model_path
+        elif isinstance(config.config, foe.ETAModelConfig) and isinstance(
+            config.config.config, etal.HasPublishedModel
+        ):
+            config.config.config.model_name = None
+            config.config.config.model_path = model_path
+        else:
+            raise ValueError(
+                "Model config must implement the %s interface"
+                % etal.HasPublishedModel
+            )
+
+    # Build model
+    return config.build()
+
+
+def _merge_config(d, kwargs):
+    for k, v in kwargs.items():
+        if k in d and isinstance(d[k], dict):
+            d[k].update(v)
+        else:
+            d[k] = v
+
+
+class ModelConfig(etal.ModelConfig):
+    """Base configuration class that encapsulates the name of a :class:`Model`
+    and an instance of its associated Config class.
+
+    Args:
+        type: the fully-qualified class name of the :class:`Model` subclass
+        config: an instance of the Config class associated with the model
+    """
+
+    pass
+
+
+class Model(etal.Model):
+    """Abstract base class for models.
+
+    This class declares the following conventions:
+
+    (a)     :meth:`Model.__init__` should take a single ``config`` argument
+            that is an instance of ``<Model>Config``
+
+    (b)     Models implement the context manager interface. This means that
+            models can optionally use context to perform any necessary setup
+            and teardown, and so any code that builds a model should use the
+            ``with`` syntax
+    """
+
+    @property
+    def has_embeddings(self):
+        """Whether this instance can generate embeddings.
+
+        This method returns ``False`` by default. Methods that can generate
+        embeddings will override this via implementing the
+        :class:`EmbeddingsMixin` interface.
+        """
+        return False
+
+    @property
+    def ragged_batches(self):
+        """True/False whether :meth:`transforms` may return tensors of
+        different sizes and therefore passing ragged lists of data to
+        :meth:`predict_all` is not allowed.
+        """
```

Comment:
If `ragged_batches == True`, then `predict_all()` does NOT accept ragged lists of data (since `np.stack()` will fail). I tweaked the docstrings to clarify.